go/arrow/datatype_fixedwidth.go

```diff
@@ -398,10 +402,15 @@ func (t *TimestampType) ClearCachedLocation() {
 // so if you change the value of TimeZone after calling this, make sure to call
 // ClearCachedLocation.
 func (t *TimestampType) GetZone() (*time.Location, error) {
+	t.mx.RLock()
 	if t.loc != nil {
+		defer t.mx.RUnlock()
 		return t.loc, nil
 	}
 
+	t.mx.RUnlock()
+	t.mx.Lock()
+	defer t.mx.Unlock()
```

Comment:
Wait... :)

A full execution of `GetZone` is possible between `RUnlock()` and `Lock()`, so it's probably a good idea to check `t.loc` after the write lock is acquired.

For a lock-free way of doing this, you can combine atomic load/stores with a writers-only mutex. Porting this C++ snippet [1] to Go:

```cpp
std::atomic<Singleton*> Singleton::m_instance;
std::mutex Singleton::m_mutex;

Singleton* Singleton::getInstance() {
    Singleton* tmp = m_instance.load();
    if (tmp == nullptr) {
        std::lock_guard<std::mutex> lock(m_mutex);
        tmp = m_instance.load();
        if (tmp == nullptr) {
            tmp = new Singleton;
            m_instance.store(tmp);
        }
    }
    return tmp;
}
```

[1] https://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/#using-c11-sequentially-consistent-atomics

Reply:
I'll add the check of `t.loc` after the write lock is acquired for now. My original attempt at this was to use `atomic.Pointer` for the loc but it made the test initialization for some of the tests annoying since you can't copy an `atomic.Pointer` and I didn't want to deal with refactoring the tests right now. 

For now, the `RWMutex` should be good enough with the added check to avoid the duplicate work. Nice catch!